### PR TITLE
[FLINK-18076][table sql / client] Sql client uses wrong class loader when parsing queries

### PR DESCRIPTION
**Pull request by leonardBang**
## What is the purpose of the change

*This pull request fix the wrong class loader in sql client(sqlParser).*

## Brief change log
- In FLINK-17728, sql client supports parser statements via sql parser, sql parser should use user's class loader which may contains user's jar from `-j` or `-l` commands

## Verifying this change

* Has existed tests and passed an e2e test in local environment.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): ( no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by dawidwys** [comment]
@leonardBang Have you verified the fix? Unfortunately it does not work.

@godfreyhe I am not exactly what is your suggested solution, so sorry if I am wrong here. However, it is not enough to wrap instance of the parser with the `wrapClassloader`. The actual call to a method that requires the classloader must be wrapped.

I think the quickest though, dirty solution would be that:
```
@Override
public Parser getSqlParser(String sessionId) {
	final ExecutionContext<?> context = getExecutionContext(sessionId);
	final TableEnvironment tableEnv = context.getTableEnvironment();
	final Parser parser = ((TableEnvironmentInternal) tableEnv).getParser();
	return new Parser() {
		@Override
		public List<Operation> parse(String statement) {
			return context.wrapClassLoader(() -> parser.parse(statement));
		}

		@Override
		public UnresolvedIdentifier parseIdentifier(String identifier) {
			return context.wrapClassLoader(() -> parser.parseIdentifier(identifier));
		}
	};
}
```

Better solutions I could think of are:
1. Do not instantiate the sources during parsing. I don't see a reason for that. IMO the logical plan could have just the CatalogTable, it does not necessary need to have an instantiated DynamicTableSource already. This is very demanding solution that requires lots of changes in the planner.
2. Wrap the call to `sqlParser.parse` in `org.apache.flink.table.client.cli.SqlCommandParser`. This is tricky as we do not have access to the user classloader there.

**Comment by godfreyhe** [comment]
@dawidwys For a long term, I think sql client should has an extended sql parser which can parse not not standard sql statement but also some special commands (e.g. `SOURCE`, `SET`, `REST`). which is also discussed in other prs. For now, some configs for both parser are different, such as: `SqlParserImplFactory`, `identifierMaxLength` (see: `PlannerContext#getSqlParserConfig` for blink planner and `PlanningConfigurationBuilder#getSqlParserConfig` for old planner). So it's also not clean that we copy those logic to sql client to build `org.apache.calcite.sql.parser.SqlParser`. As an interim solution, we use public interface `org.apache.flink.table.delegation.Parser` to parse a sql statement, which not only parse a statement but also validate the SqlNode (will initialize `DynamicTableSource`).

**Comment by leonardBang** [comment]
@dawidwys Sorry for lost my first commit, You’re right my second commit can not fix this, could you have a look  my first commit(https://github.com/apache/flink/commit/41240cd524255ffc7cf63d2bc2ee9ff38f3abcbd) ?
I'll push it again, I revert this the first one because they expose ExecutionContext from Executor which we think it's not a good way to fix after communicate with @godfreyhe

**Comment by dawidwys** [comment]
Sure, just let me know when it's ready for review.

Honestly I also don't like too much exposing the `ExecutionContext`.

**Comment by leonardBang** [comment]
> Sure, just let me know when it's ready for review.
> 
> Honestly I also don't like too much exposing the `ExecutionContext`.
I‘ll update the PR soon

**Comment by leonardBang** [comment]
The PR updated @dawidwys , please help take a look.

File: flink-table/flink-sql-client/src/test/java/org/apache/flink/table/client/gateway/local/DependencyTest.java [hunk 1]
```diff
@@ -82,6 +87,44 @@
...
+		final SessionContext session = new SessionContext("test-session", new Environment());
+		String sessionId = executor.openSession(session);
+		try {
+			String expectedPlan = "LogicalProject(IntegerField1=[$0], StringField1=[$1])\n" +
+				"  LogicalTableScan(table=[[default_catalog, default_database, TableNumber1, source: [TestTableSource]]])\n";
+			final Parser sqlParser = executor.getSqlParser(sessionId);
+			List<Operation> operations = sqlParser.parse("SELECT IntegerField1, StringField1 FROM TableNumber1");
+
+			assertTrue(operations != null && operations.size() == 1);
+			String actualPlan = RelOptUtil.toString(((PlannerQueryOperation) operations.get(0)).getCalciteTree());
```
Comment (godfreyhe) [inline flink-table/flink-sql-client/src/test/java/org/apache/flink/table/client/gateway/local/DependencyTest.java:119]: I think plan validation is not necessary, only checking parser is enough.

**Comment by leonardBang** [comment]
Thanks for the review, I updated the PR @dawidwys @godfreyhe

**Comment by dawidwys** [comment]
Thank you both, will merge once azure gives green.

**Pull request closed without merging**